src/components/SelectionPills/SelectionPills.js

```diff
@@ -0,0 +1,196 @@
+    label: PropTypes.string,
+
+    /** Option to only allow a maximum number of selected items. No restriction if not set. */
+    maxSelectionCount: PropTypes.number,
+
+    /** Option to include a generated pill that will toggle all other pills on / off. Disabled if a maxSelectionCount or parentControlledState. */
+    includeSelectAll: PropTypes.bool,
+
+    /** Optional override of the select all pill label */
+    selectAllLabel: PropTypes.string,
+
+    /** Optional override styles */
+    style: PropTypes.shape({
+      wrapperStyle: PropTypes.object,
+      listStyle: PropTypes.object,
+      labelStyle: PropTypes.object,
+    }),
+  }
+
+  static defaultProps = {
+    elementAttributes: {},
+    onSelectPill: NoOp,
+    pills: [],
+    maxSelectionCount: null,
+    includeSelectAll: false,
+    selectAllLabel: 'All',
+    parentControlledState: false,
+    style: {},
+  }
+
+  state = {
+    pillsList: this.props.pills,
+    selectedCount: this.props.pills.filter(p => p.isSelected === true).length,
+    allSelected: false,
+  }
+
+  // Initialize pills selected states
+  componentDidMount() {
+    this.setState(prevState => ({
+      pillsList: prevState.pillsList.map(p => {
+        return {
+          ...p,
+          isSelected: p.isSelected || false,
+        }
+      }),
+    }))
+  }
+
+  onSelectAll = e => {
+    const { selectAllLabel, onSelectPill } = this.props
+    const { pillsList, allSelected } = this.state
+    const selectedPill = { id: 'selectAllPill', isSelected: !allSelected, text: selectAllLabel }
+    const newPillList = pillsList.map(p => {
+      return {
+        ...p,
+        isSelected: false,
+      }
+    })
+
+    this.setState({
+      pillsList: newPillList,
+      allSelected: !allSelected,
+    })
+    onSelectPill(e, selectedPill, newPillList)
+  }
+
+  onSelectPill = (e, pill) => {
+    const { onSelectPill, pills, parentControlledState } = this.props
+    const selectedPill = { ...pill, isSelected: !pill.isSelected }
+    if (parentControlledState) {
+      return onSelectPill(e, selectedPill, pills)
+    }
+
+    const { pillsList, selectedCount } = this.state
+    const newPillList = pillsList.map(p => {
+      if (p.id !== pill.id) {
+        return p
+      }
+      return selectedPill
+    })
+
+    this.setState({
+      pillsList: newPillList,
+      selectedCount: pill.isSelected ? selectedCount - 1 : selectedCount + 1,
+      allSelected: false,
+    })
+    onSelectPill(e, selectedPill, newPillList)
+  }
+
+  // Disable not selected pills when max pills reached. Not applicable when parent controlled
+  isDisabledPill = pill => {
+    const { maxSelectionCount, parentControlledState } = this.props
+    const { selectedCount } = this.state
+    if (parentControlledState || pill.isSelected) return false
+    if (!selectedCount || !maxSelectionCount) return false
+    return selectedCount >= maxSelectionCount
+  }
+
+  renderSelectAllPill = () => {
```

Comment:
More into design system territory, any thoughts on making the select all pill optional? This is something I'd use for a feature I'm working on, and there's a good chance I'd want to leave out a 'select all' option.

Reply:
The select all is totally optional and defaulted to false. If you look at the gifs I provided, the first gif shows the default behavior and does not include a select all. It's controlled through the `includeSelectAll` prop